File: app/src/main/java/com/dingyi/visualizer/drawable/CircleDrawable.java
package com.dingyi.visualizer.drawable;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.PointF;

import com.dingyi.visualizer.util.LogUtil;
import com.dingyi.visualizer.util.MathUtil;

import java.util.Arrays;
import java.util.Collections;

public class CircleDrawable extends BaseDrawable {


    private float[] linesArray;

    private float[] pointsArray;

    private float cx;

    private float cy;

    private float baseCr;

    private float cr;

    private final Paint circlePaint = new Paint();

    private final Paint arcPaint = new Paint();

    private final PointF circlePointF = new PointF();

    private final PointF newCirclePointF = new PointF();

    private final PointF twoCirclePointF = new PointF();


    public CircleDrawable(Context context) {
        super(context);
    }


    @Override
    public void init() {

        //circle paint
        circlePaint.setColor(getDataBean().getColorData().getCirclePaintColor());
        circlePaint.setStrokeWidth(getDataBean().getColorData().getCirclePaintWidth());
        circlePaint.setStrokeCap(Paint.Cap.ROUND);
        circlePaint.setStyle(Paint.Style.FILL);

        //arc paint
        arcPaint.setColor(getDataBean().getColorData().getArcPaintColor());
        arcPaint.setColor(getDataBean().getColorData().getCirclePaintColor());
        arcPaint.setStrokeCap(Paint.Cap.ROUND);
        arcPaint.setStyle(Paint.Style.STROKE);
    }



    @Override
    public void onDraw(Canvas canvas) {
        if (cy == 0) {
            readyData(canvas);
        }

        System.gc();

       for (int i = 0; i < getDataBean().getCircleLineCount(); i = i + 1) {

            MathUtil.getCirclePoint(
                    cx, cy, cr + dp2px(3), i * getDataBean().getCircleAngle()+270, circlePointF);


            float des = (float) getData()[getDataBean().getSkipStartArrayIndex()+i];


            MathUtil.getPoint(
                    cx, cy,
                    MathUtil.getPointDistance(cx, cy, circlePointF.x, circlePointF.y) + des,
                    MathUtil.getPointAngle(cx, cy, circlePointF.x, circlePointF.y), newCirclePointF);

            linesArray[i * 4] = circlePointF.x;
            linesArray[i * 4 + 1] = circlePointF.y;
            linesArray[i * 4 + 2] = newCirclePointF.x;
            linesArray[i * 4 + 3] = newCirclePointF.y;


            MathUtil.getPoint(
                    cx, cy,
                    MathUtil.getPointDistance(cx, cy, circlePointF.x, circlePointF.y) - des - dp2px(4),
                    MathUtil.getPointAngle(cx, cy, circlePointF.x, circlePointF.y), twoCirclePointF);

            pointsArray[i * 2] = twoCirclePointF.x;
            pointsArray[i * 2 + 1] = twoCirclePointF.y;


        }


        canvas.drawArc(cx - cr, cy - cr, cx + cr, cy + cr, 0, 360, false, arcPaint);
        canvas.drawLines(linesArray, circlePaint);
        if (getDataBean().isShowCirclePoint()){
            canvas.drawPoints(pointsArray, circlePaint);
        }


    }

    private void readyData(Canvas canvas) {
        linesArray = new float[(int) getDataBean().getCircleLineCount() * 4];
        pointsArray = new float[(int) getDataBean().getCircleLineCount() * 2];
        cx = canvas.getWidth() / 2;
        cy = canvas.getHeight() / 2;
        baseCr = getDataBean().getCircleRadius();
        cr = baseCr;
    }


}

File: app/src/main/kotlin/io/github/dingyi222666/visualizer_app/drawable/CircleDrawable.kt
package io.github.dingyi222666.visualizer_app.drawable

import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.PointF
import io.github.dingyi222666.visualizer_app.utils.MathUtil
import io.github.dingyi222666.visualizer_app.visualizer.VisualizerConfig

/**
 * Circle-based audio visualization drawable
 */
class CircleDrawable(context: Context) : BaseDrawable(context) {
    
    // Arrays for drawing
    private var linesArray: FloatArray? = null
    private var pointsArray: FloatArray? = null
    
    // Canvas dimensions
    private var centerX: Float = 0f
    private var centerY: Float = 0f
    private var baseRadius: Float = 0f
    private var radius: Float = 0f
    
    // Rotation variables
    private var rotationAngle: Float = 0f
    private val rotationSpeed: Float = 0.01f
    
    // Paint objects
    private val circlePaint = Paint()
    private val arcPaint = Paint()
    
    // Reusable point objects for calculations
    private val circlePointF = PointF()
    private val newCirclePointF = PointF()
    private val twoCirclePointF = PointF()
    
    /**
     * Initialize the drawable with default settings
     */
    override fun init() {
        val config = getConfig() ?: return
        
        // Circle paint for lines and points
        circlePaint.apply {
            color = config.colors.primaryColor
            strokeWidth = config.colors.circleLineWidth
            strokeCap = Paint.Cap.ROUND
            style = Paint.Style.FILL
        }
        
        // Arc paint for the center circle
        arcPaint.apply {
            color = config.colors.circleOutlineColor
            strokeCap = Paint.Cap.ROUND
            style = Paint.Style.STROKE
            strokeWidth = config.colors.circleOutlineWidth
        }
    }
    
    /**
     * Called when configuration changes
     */
    override fun onConfigChanged(config: VisualizerConfig) {
        // Reinitialize paint objects with new colors/widths
        init()
    }
    
    /**
     * Initialize data arrays and dimensions based on canvas size
     */
    private fun prepareDrawing(canvas: Canvas) {
        val config = getConfig() ?: return
        
        val lineCount = config.circle.lineCount.toInt()
        
        // Create arrays if needed or if size has changed
        if (linesArray == null || linesArray!!.size != lineCount * 4) {
            linesArray = FloatArray(lineCount * 4)
        }
        
        if (pointsArray == null || pointsArray!!.size != lineCount * 2) {
            pointsArray = FloatArray(lineCount * 2)
        }
        
        // Calculate center and radius based on canvas dimensions
        centerX = canvas.width / 2f
        centerY = canvas.height / 2f
        baseRadius = config.circle.radius
        radius = baseRadius
    }
    
    /**
     * Draw the visualization
     */
    override fun onDraw(canvas: Canvas) {
        val config = getConfig() ?: return
        val data = getData() ?: return
        
        // Initialize dimensions if needed
        if (centerY == 0f) {
            prepareDrawing(canvas)
        }
        
        // Update rotation angle
        rotationAngle = (rotationAngle + rotationSpeed) % 360
        
        val skipStartIndex = config.dataStartIndex
        val lineCount = config.circle.lineCount.toInt()
        val angle = config.circle.angle
        val lines = linesArray ?: return
        val points = pointsArray ?: return
        
        // Calculate points for lines and dots
        for (i in 0 until lineCount) {
            // Get point on the base circle with rotation
            MathUtil.getCirclePoint(
                centerX, centerY, radius + 3.px,
                i * angle + 270 + rotationAngle, circlePointF
            )
            
            // Get amplitude at this position
            val amplitude = if (skipStartIndex + i < data.size) {
                data[skipStartIndex + i].toFloat()
            } else {
                0f
            }
            
            // Calculate outer point (extending outward from the circle)
            MathUtil.getPoint(
                centerX, centerY,
                MathUtil.getPointDistance(centerX, centerY, circlePointF.x, circlePointF.y) + amplitude,
                MathUtil.getPointAngle(centerX, centerY, circlePointF.x, circlePointF.y), 
                newCirclePointF
            )
            
            // Calculate inner point (extending inward from the circle)
            MathUtil.getPoint(
                centerX, centerY,
                MathUtil.getPointDistance(centerX, centerY, circlePointF.x, circlePointF.y) - amplitude - 4.px,
                MathUtil.getPointAngle(centerX, centerY, circlePointF.x, circlePointF.y), 
                twoCirclePointF
            )
            
            // Store line coordinates (from circle point to outer point)
            lines[i * 4] = circlePointF.x
            lines[i * 4 + 1] = circlePointF.y
            lines[i * 4 + 2] = newCirclePointF.x
            lines[i * 4 + 3] = newCirclePointF.y
            
            // Store inner point coordinates
            points[i * 2] = twoCirclePointF.x
            points[i * 2 + 1] = twoCirclePointF.y
        }
        
        // Draw the base circle
        canvas.drawArc(
            centerX - radius, 
            centerY - radius, 
            centerX + radius, 
            centerY + radius, 
            0f, 360f, false, arcPaint
        )
        
        // Draw the radial lines
        canvas.drawLines(lines, circlePaint)
        
        // Draw the inner points if enabled
        if (config.circle.showPoints) {
            canvas.drawPoints(points, circlePaint)
        }
    }
    
    /**
     * Release resources
     */
    override fun release() {
        super.release()
        linesArray = null
        pointsArray = null
    }
} 